src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+        {
+            this.logger.LogTrace("()");
+
+            if (this.storeSettings.ReIndex)
+                throw new NotImplementedException();
+
+            ChainedHeader initializationTip = this.chain.GetBlock(this.blockRepository.BlockHash);
+            this.SetStoreTip(initializationTip);
+            
+            if (this.StoreTip == null)
+                await this.RecoverStoreTipAsync().ConfigureAwait(false);
+
+            this.logger.LogDebug("Initialized block store tip at '{0}'.", this.StoreTip);
+
+            if (this.storeSettings.TxIndex != this.blockRepository.TxIndex)
+            {
+                if (this.StoreTip != this.chain.Genesis)
+                    throw new BlockStoreException("You need to rebuild the block store database using -reindex-chainstate to change -txindex");
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Make sure that block store initializes before the consensus. 
+            // This is needed in order to rewind consensus in case it is initialized ahead of the block store.
+            if (this.chainState.ConsensusTip != null)
+            {
+                this.logger.LogCritical("Block store initialized after the consensus!");
+                throw new BlockStoreException("Block store initialized after consensus!");
+            }
+
+            // Start dequeuing.
+            this.currentBatchSizeBytes = 0;
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the store tip.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.StoreTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
+        }
+
+        /// <summary>
+        /// Sets block store tip to the last block that exists both in the repository and in the <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task RecoverStoreTipAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var blockStoreResetList = new List<uint256>();
+            Block resetBlock = await this.blockRepository.GetAsync(this.blockRepository.BlockHash).ConfigureAwait(false);
+            uint256 resetBlockHash = resetBlock.GetHash();
+
+            while (this.chain.GetBlock(resetBlockHash) == null)
+            {
+                blockStoreResetList.Add(resetBlockHash);
+
+                if (resetBlock.Header.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    resetBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+                Guard.NotNull(resetBlock, nameof(resetBlock));
```

Comment:
Seems unusual to have a mid method guard like this. Could it be better to either ensure `this.blockRepository.GetAsync` never returns null? Perhaps by a proper exception inside of that rather than returning null and losing that context?

Reply:
>Could it be better to ensure this.blockRepository.GetAsync never returns null?

No. It is supposed to return null if there is no block that correspond to a given hash. 
So guard here and not on lower level is intentional.

Reply:
When you say "supposed to" - why is that? It might be a better design rather than relying on the client code to do a null check, and in this case throw exception.  Though this is more around the design of the block repository rather than the design of this calling code. So happy to let this one go for now.

Reply:
>When you say "supposed to" - why is that?

Because that's what that method does. 
Even if I did want to throw if it doesn't exist- changing the repository logic is well outside of the scope of this pr. And it also requires checking every one of 20 usages of that method. 

If you feel that repository should throw exception when there is no block- please do raise an issue.